sw/host/opentitanlib/src/spiflash/flash.rs

```diff
@@ -0,0 +1,282 @@
+        let mut buf = vec![0u8; 256];
+        let mut tries = 0;
+        loop {
+            spi.run_transaction(&mut [
+                // READ_SFDP always takes a 3-byte address followed by a dummy
+                // byte regardless of address mode.  To simplify, we always
+                // read from address zero.
+                Transfer::Write(&[SpiFlash::READ_SFDP, 0, 0, 0, 0]),
+                Transfer::Read(&mut buf),
+            ])?;
+
+            // We only want to give SFDP parsing one extra chance for length
+            // extension. If parsing fails a second time, just return the error.
+            let result = Sfdp::try_from(&buf[..]);
+            if result.is_ok() || tries > 0 {
+                return Ok(result?);
+            }
+            buf.resize(Sfdp::length_required(&buf)?, 0);
+            tries += 1;
+        }
+    }
+
+    /// Create a new `SpiFlash` instance from an SFDP table.
+    pub fn from_sfdp(sfdp: Sfdp) -> Self {
+        let (erase_sz, erase_op) = match sfdp.jedec.block_erase_size {
+            BlockEraseSize::Block4KiB => (4096, sfdp.jedec.erase_opcode_4kib),
+            _ => (sfdp.jedec.sector[0].size, sfdp.jedec.sector[0].erase_opcode),
+        };
+        SpiFlash {
+            size: sfdp.jedec.density,
+            erase_size: erase_sz,
+            erase_opcode: erase_op,
+            program_size: SpiFlash::LEGACY_PAGE_SIZE,
+            address_mode: AddressMode::from(sfdp.jedec.address_modes),
+            sfdp: Some(sfdp),
+        }
+    }
+
+    /// Create a new `SpiFlash` instance by reading an SFDP table from the `spi` Target.
+    pub fn from_spi(spi: &dyn Target) -> Result<Self> {
+        let sfdp = SpiFlash::read_sfdp(spi)?;
+        Ok(SpiFlash::from_sfdp(sfdp))
+    }
+
+    /// Set the SPI flash addressing mode to either 3b or 4b mode.
+    pub fn set_address_mode(&mut self, spi: &dyn Target, mode: AddressMode) -> Result<()> {
+        let opcode = [match mode {
+            AddressMode::Mode3b => SpiFlash::EXIT_4B,
+            AddressMode::Mode4b => SpiFlash::ENTER_4B,
+        }];
+        spi.run_transaction(&mut [Transfer::Write(&opcode)])?;
+        self.address_mode = mode;
+        Ok(())
+    }
+
+    /// Automatically set the addressing mode based on the size of the SPI flash.
+    pub fn set_address_mode_auto(&mut self, spi: &dyn Target) -> Result<()> {
+        self.set_address_mode(
+            spi,
+            if self.size <= 16 * 1024 * 1024 {
+                AddressMode::Mode3b
+            } else {
+                AddressMode::Mode4b
+            },
+        )
+    }
+
+    /// Read into `buffer` from the SPI flash starting at `address`.
+    pub fn read(&self, spi: &dyn Target, mut address: u32, buffer: &mut [u8]) -> Result<()> {
+        let mut remain = buffer.len();
+        let mut chunkstart = 0usize;
+        while remain != 0 {
+            // Break the read up according to the maximum chunksize the backend can handle.
```

Comment:
Maybe you can do something with `<[T]>::chunks()`?

Reply:
Done